lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,425 @@
+use crate::{gelf_fields::*, VALID_FIELD_REGEX};
+use bytes::{BufMut, BytesMut};
+use serde::{Deserialize, Serialize};
+use tokio_util::codec::Encoder;
+use value::Value;
+use vector_core::{
+    config::{log_schema, DataType},
+    event::Event,
+    event::LogEvent,
+    schema,
+};
+
+/// On GELF encoding behavior:
+///   Graylog has a relaxed parsing. They are much more lenient than the spec would
+///   suggest. We've elected to take a more strict approach to maintain backwards compatability
+///   in the event that we need to change the behavior to be more relaxed, so that prior versions
+///   of vector will still work.
+///   The exception is that if 'Additional fields' are found to be missing an underscore prefix and
+///   are otherwise valid field names, we prepend the underscore.
+
+static MISSING_FIELD_STR: &str = "LogEvent does not contain required field.";
+
+static INVALID_TYPE_STR: &str = "LogEvent contains a value with an invalid type.";
+
+static INVALID_FIELD_NAME_STR: &str = "LogEvent contains a value with an invalid type.";
+
+/// Config used to build a `GelfSerializer`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct GelfSerializerConfig;
+
+impl GelfSerializerConfig {
+    /// Creates a new `GelfSerializerConfig`.
+    pub const fn new() -> Self {
+        Self
+    }
+
+    /// Build the `GelfSerializer` from this configuration.
+    pub fn build(&self) -> GelfSerializer {
+        GelfSerializer::new()
+    }
+
+    /// The data type of events that are accepted by `GelfSerializer`.
+    pub fn input_type() -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement() -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // JSON, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the GELF format.
+/// Spec: https://docs.graylog.org/docs/gelf
+#[derive(Debug, Clone)]
+pub struct GelfSerializer;
+
+impl GelfSerializer {
+    /// Creates a new `GelfSerializer`.
+    pub fn new() -> Self {
+        GelfSerializer
+    }
+
+    /// Encode event and represent it as JSON value.
+    pub fn to_json_value(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
+        match event {
+            Event::Log(log) => serde_json::to_value(&log),
+            Event::Metric(_) | Event::Trace(_) => {
+                panic!("GELF Serializer does not support Metric or Trace events.")
+            }
+        }
+    }
```

Comment:
Oh... I think I made this mistake because the other Serializers that supported this were already encoded in json.

It seems then that perhaps the return value of this function should be:

```Rust
Result<serde_json::Value, vector_core::Error>
```

? Otherwise, would have to construct a serde_json Error if the encoding fails.

Reply:
Makes sense, feel free to change the signatures where fit!

(You can construct a `serde_json::Error` by calling [`serde_json::Error::custom`](https://docs.serde.rs/serde/de/trait.Error.html#tymethod.custom), but I agree that it doesn't necessarily fit well here.)

Reply:
Will do- I see now the comment in the method is pretty clear, I just overlooked it 😬  "Encode event ..."

Reply:
Attempted to address in latest commit.